torch/utils/_sympy/value_ranges.py

```diff
@@ -323,46 +328,48 @@ def floordiv(a, b):
         else:
             return ValueRanges.coordinatewise_monotone_map(a, b, operator.floordiv)
 
-    @staticmethod
-    def truncdiv(a, b):
-        a = ValueRanges.wrap(a)
-        b = ValueRanges.wrap(b)
-        if 0 in b or ((-sympy.oo in a or sympy.oo in a) and (-sympy.oo in b or sympy.oo in b)):
-            return ValueRanges.unknown()
-        else:
-            # Casting to integer does truncation
-            def f(a, b):
-                result = a / b
-                # This won't work for sympy.Expr, so it'll need a workaround when
-                # dealing with dynamic shapes
-                if result.is_finite:
-                    result = sympy.Integer(result)
-                return result
-            return ValueRanges.coordinatewise_monotone_map(a, b, f)
+    @classmethod
+    def truncdiv(cls, a, b):
+        x = cls.truediv(a, b)
+        if x == ValueRanges.unknown():
+            return x
 
-    @staticmethod
-    def div(a, b):
-        return ValueRangeAnalysis.truediv(a, b)
+        def trunc(x):
+            return sympy.Integer(x) if x.is_finite else x
+
+        return ValueRanges.increasing_map(x, trunc)
+
+    @classmethod
+    def div(cls, a, b):
+        return cls.truediv(a, b)
 
     @staticmethod
     def add(a, b):
         return ValueRanges.coordinatewise_increasing_map(a, b, operator.add)
 
-    @staticmethod
-    def mul(a, b):
+    @classmethod
+    def mul(cls, a, b):
+        a = ValueRanges.wrap(a)
+        b = ValueRanges.wrap(b)
+
+        assert a.is_bool == b.is_bool
+        if a.is_bool:
+            return cls.and_(a, b)
+
         def safe_mul(a, b):
+            # Make unknown() * wrap(0) == wrap(0)
             if a == 0:
-                return 0
+                return a
```

Comment:
jc, why did you have to return a here instead of 0?

Reply:
to return a zero that's both a sympy object and it's of the correct dtype.